delivery_carrier_account/models/carrier_account.py

```diff
@@ -0,0 +1,44 @@
+# Copyright 2014-TODAY Florian da Costa Akretion <http://www.akretion.com>.
+# Copyright 2014 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+
+
+class CarrierAccount(models.Model):
+    _name = "carrier.account"
+    _description = "Base account datas"
+    _order = "sequence"
+
+    @api.model
+    def _selection_file_format(self):
+        """To inherit to add label file types"""
+        return [("PDF", "PDF"), ("ZPL", "ZPL"), ("XML", "XML")]
+
+    name = fields.Char(required=True)
+    sequence = fields.Integer()
+    delivery_type = fields.Selection(
+        selection=lambda self: self.env["delivery.carrier"]
+        ._fields["delivery_type"]
+        .selection,
+        help="This field may be used to link an account to a carrier",
+    )
+    carrier_ids = fields.Many2many(
```

Comment:
I don't like the complication on the design having a m2m instead of a many2one is worth. I know the case you have exposed, but this is an anti-UX design, and also not compatible with the rest of the system (you won't see which specific delivery carrier case will be used in for example the sales order). It's not too-much trade-off to create 2 delivery carriers linked to the same account.

Reply:
This many2many is totally optional and should be used only for specific carriers.
I don't see how it can be replaced by a many2one though.

The most common case is to have one account by carrier, and possibly multiple delivery method by carrier.
In that case, you don't need to to fill this `carrier_ids` field, because the `delivery_type` you chose is enough to find the right account.
But in the case for a same carrier, you may have to choose a different account depending on the delivery method, then you have to fill the field, and only in this case.
It happens for DPD (different account for the relay point delivery method) and Geodis for example.

Since the case may happen, I think it is important to keep the possibility to link an account to delivery method this way, but what could be done is to show this field only for carrier that may need it.
For instance, adding an invisible computed field `show_carrier_ids` which would just check if a method `_get_delivery_type_allowing_carrier_ids` (I'd have to find a better name for the method..) contains the carrier account's delivery type.

Then, the modules delivery_dpd, delivery_geodis could just override `_get_delivery_type_allowing_carrier_ids` to add their own `delivery_type` and make the field visible.
Would it be better this way?

Reply:
The anti-pattern I'm seeing is that you select the match on the account through delivery_type, instead of selecting the account on the delivery carrier. That is a lot more confusing. So I think we should put a many2one `delivery_account_id` in `delivery.carrier` for selecting the account, restricting the possible values to the ones with the same `delivery_type`.

Reply:
But one of the goal of this module beeing to avoid a lot of data duplication but having the possibility to have multiple account for 1 delivery carrier, the many2one can't work.

The delivery type beeing a carrier and the delivery method beeing just a specific service of the carrier, it makes sense that we select the match on the account throught delivery type.
And we still keep possibility to filter by delivery method, only if needed, to keep some flexibility.

I can understand the preference to configure and show a direct link on the delivery method, but we need to cover all the uses cases and this does not...

Another idea could be to add a computed readonly m2o field on the side on the delivery_carrier, to show which account will be chosen for a given delivery method, so it is very clear of the users. But the field would only be informational.
And, if there are multiple accounts for a delivery order, which is a more advanced case, then it would stay empty.

I am really open to changes, but the uses cases stated must continue to be possible.

Reply:
As said, adding several carrier accounts is an anti-pattern that makes you later to have to select the proper one on another step, complicating everything, and not being able to show the proper operation on the right moment (for example, on the sales order). The use case should be covered with several carriers, not with one.

Reply:
@pedrobaeza Ok, clear enough.
Well, on my side, I am doing this migration for the customer that needs multiple account per carrier/delivery method (white label dropshipping case explained before).
They use 5 carriers/23 delivery methods / 5 accounts by carrier (5 white label company as customer).

If we add a many2one on the `delivery.carrier`, it will indeed make the multiple accounts per delivery method impossible, which was one of the initial goal of this module (model). It would force me to configure 115 `delivery.method` (instead of 23).
Probably even more in the future.
I won't use the module if it is based on a many2one on `delivery.carrier`.
Sadly, I see 2 options only : 
1) We keep the module as it is, since it is already a used OCA module from several version
I understand you'd continue not using it
2) Merge is blocked into the OCA until the change you ask is done => It won't be usable for me anymore, I'll use a fork or something and will leave someone else continue on this PR.

Don't see any lack of willingness in the proposal, I am just no seeing other solution.

Reply:
A third option is to keep it with the many2one, and just add the m2m in a module over it with such functionality calling it `delivery_carrier_account_multi_carrier`.

Reply:
I still don't see how do the user selects the proper account in the picking or whatever? Or do you tell me that it only relies on the method `_get_carrier_account` to return the proper one because the user knows how to avoid collisions in the configuration?

I still see a very bad idea to do this m2m.

Reply:
It only relies on `_get_carrier_account` The normal user don't need to know which account is used, it does not matter to him

Carrier accounts basically never changes, the person in charge of its configuration have to get the information from the carrier, if for 1 carrier, there are multiple accounts, the person in charge is very well aware of the reason and he configures the account, once when the carrier start beeing used.
Basic case, you just create your 1 account with one delivery type, nothing difficult.
less basic case, you have different accounts depending on the delivery.method, you have to select the delivery type, and then since you know that you have multiple account for the carrier, you also select the delivery methods concerned.
Advanced case, you have multiple account for a delivery method, it depends on other criterias (in my case a sale channel)
The person know he has 1 account by sale channel, he goes and select delivery type and the sale  channel concerned (require additional module of course). In my experience, it never has been an issue, and, once it is done, it never changes. 

Anyway, I understand you don't deal with what I called "advanced case" so you don't see why we would not just have a many2one on carrier which would indeed be the easiest. But we have to leave a solution for advanced case IMO.
Like I said, without this possibility, I just won't have any use for the module for example.

In order to go forward with this subject, I propose we then go with this imperfect solution : 
Remove `carrier_ids` and remove `_get_carrier_account_domain`
Add the many2one on `delivery.carrier` (I guess mandatory if delivery_type is not fixed/based on rule)
Adapt `_get_carrier_account` to return the account set on the `delivery.carrier`

Then, each carrier specific modules must use the method `_get_carrier_account`.
If some module depends on `delivery_carrier_account` and don't use `_get_carrier_account`, we'll just make a PR to change that.
Not perfect because I guess some module won't use the method, but, in case someone needs it to call the method, it is an easy/fast PR.

For the advanced case management, it will be enough to : 
Hide the many2one on `delivery.carrier`
Add the required fields to manage multiple account on `carrier.account`
Override `_get_carrier_account` to call a new method `_get_carrier_account_domain`

It is quite easy, it does not complexify `carrier_account` 
The only issue is that we can't make carrier_specific module to use the `_get_carrier_account` method but I will add a comment and if it is a problem, we'll just adapt the specific carrier modules when the problem occurs.

What do you think ?

Reply:
I have made the changes, if you want to have a look.
I have added a dependency on web_domain_field, to manage the domain on the many2one on the `delivery.carrier `model.
Not sure we have another choice to manage the case, but the module since well maintained so it seems quite reasonable.